plugins/UM3NetworkPrinting/src/SendMaterialJob.py

```diff
@@ -1,99 +1,193 @@
+        Logger.log("d", "Syncing material {material_id} with cluster.".format(material_id = material_id))
+        self.device.postFormWithParts(target = "materials/", parts = parts, on_finished = self.sendingFinished)
 
-    def sendingFinished(self, reply: QNetworkReply):
+    ##  Check a reply from an upload to the printer and log an error when the call failed
+    @staticmethod
+    def sendingFinished(reply: QNetworkReply) -> None:
         if reply.attribute(QNetworkRequest.HttpStatusCodeAttribute) != 200:
-            Logger.log("e", "Received error code from printer when syncing material: {code}".format(code = reply.attribute(QNetworkRequest.HttpStatusCodeAttribute)))
-            Logger.log("e", reply.readAll().data().decode("utf-8"))
\ No newline at end of file
+            Logger.log("e", "Received error code from printer when syncing material: {code}, {text}".format(
+                code = reply.attribute(QNetworkRequest.HttpStatusCodeAttribute),
+                text = reply.errorString()
+            ))
+
+    ##  Parse the reply from the printer
+    #
+    #   Parses the reply to a "/materials" request to the printer
+    #
+    #   \return a dictionary of ClusterMaterial objects by GUID
+    #   \throw KeyError Raised when on of the materials does not include a valid guid
+    @classmethod
+    def _parseReply(cls, reply: QNetworkReply) -> Dict[str, ClusterMaterial]:
+        try:
+            remote_materials = json.loads(reply.readAll().data().decode("utf-8"))
+            return {material["guid"]: ClusterMaterial(**material) for material in remote_materials}
+        except UnicodeDecodeError:
+            Logger.log("e", "Request material storage on printer: I didn't understand the printer's answer.")
+        except json.JSONDecodeError:
+            Logger.log("e", "Request material storage on printer: I didn't understand the printer's answer.")
+        except TypeError:
+            Logger.log("e", "Request material storage on printer: Printer's answer was missing GUIDs.")
+
+    ##  Retrieves a list of local materials
+    #
+    #   Only the new newest version of the local materials is returned
+    #
+    #   \return a dictionary of LocalMaterial objects by GUID
+    def _getLocalMaterials(self) -> Dict[str, LocalMaterial]:
+        result = {}  # type: Dict[str, LocalMaterial]
+        container_registry = Application.getInstance().getContainerRegistry()
+        material_containers = container_registry.findContainersMetadata(type = "material")
+
+        # Find the latest version of all material containers in the registry.
+        for material in material_containers:
+            try:
+                # material version must be an int
+                material["version"] = int(material["version"])
+
+                # Create a new local material
+                local_material = LocalMaterial(**material)
```

Comment:
Populating the `namedtuple` with unknown keys will throw a `TypeError`, we should make this a bit more robust (as containers from the registry can contain all sorts of data).